src/controllers/schedule.js

```diff
@@ -0,0 +1,281 @@
+        const [tasks, task_err] = await taskModel.getTasksForScheduling(user_id, req.body.ids);
+        if (task_err != null) {
+            return res.status(400).json({
+                data: null,
+                error: "No unscheduled tasks"
+            });
+        }
+
+        console.log(tasks);
+
+        // Create the freeBusy request
+        const [user, _] = await userModel.getUser(user_id);     
+        const relevantCalendars = user.relevant_calendars.split(',');
+        const freeBusyItems = [];
+
+        relevantCalendars.forEach((calendarID) => {
+            freeBusyItems.push({"id": calendarID});
+        });
+
+        const freeBusyBody = {
+            "timeMin": req.body.timeMin, // in UTC
+            "timeMax": tasks[tasks.length - 1].due_date, // in UTC
+            "timeZone": req.body.timeZone, // user's local timezone
+            "items": freeBusyItems
+        };
+
+        console.log(freeBusyBody);
+
+        // Get the list of busy intervals in user's local time
+        const [googleBusyInts, int_err] = await googleAPIService.getFreeBusyIntervalsWithToken(req.headers, freeBusyBody);
+    
+        // Sort the busy intervals in ascending order of start datetime
+        googleBusyInts.sort((a, b) => {
+            if (moment(a[0]).isBefore(moment(b[0])))
+                return -1;
+            else if (moment(a[0]).isAfter(moment(b[0])))
+                return 1;
+            else
+                return 0;
+        });
+
+        // Convert the busy intervals to moments
+        const busyMomInts = googleBusyInts.map((interval) => {
+            return [
+                moment(interval[0]).tz(req.body.timeZone), 
+                moment(interval[1]).tz(req.body.timeZone)
+            ];
+        });
+
+        // Print for debugging purposes
+        console.log("BUSY INTERVALS");
+        busyMomInts.forEach((busyInt) => {
+            console.log([busyInt[0].format(), busyInt[1].format()]);
+        });
+        console.log();
+
+        // Get user's hours of operation (should be in 24hr format)
+        const startHr = user.hours_start;
+        const endHr = user.hours_end;
+
+        // Expects a moment object
+        function validTime(momentObj) {
+            if (momentObj.hour() >= startHr && momentObj.hour() < endHr) {
+                console.log('within hours of operation');
+                return 0;
+            }
+            else if (momentObj.hour() < startHr) {
+                console.log('before hours of operation');
+                return -1;
+            }
+            else if (momentObj.hour() >= endHr) {
+                console.log('after hours of operation');
+                return 1;   
+            }
+            console.log('what');
+        }
+
+        // Create free intervals from busy intervals in user's local time
+        let freeInts = [];
+        let lastEndTime = moment(req.body.timeMin).tz(req.body.timeZone);
+        busyMomInts.forEach((busyInt) => {
```

Comment:
This should go into `ScheduleService.js` -- the task here is convert busy times to free times, using user's hours of operation. You can make busy times and hours of operation all parameters to the function so `ScheduleService.js` does not have to fetch the user model again